### DISCUSSION TITLE
How to get HTTP headers in DataFetcher?

### DISCUSSION THREAD

**Discussion by powerseed** [graphql-java/graphql-java#2583] [Q&A] [answered]
For example, as shown below, I want to send a `cf-ipcountry` attribute as HTTP header to my Java BE project using GraphQL Playground.

![image](https://user-images.githubusercontent.com/42278237/137236900-c7964c64-9919-4ec6-a0df-ac384561fb34.png)

I am wondering how to get this `cf-ipcountry` attribute in a DataFetcher?

My DataFetcher is like below
```
    public ConnectedDataFetcher<String> getCurrency() {
        return ConnectedDataFetcher.create(this.txManager, env -> {
            User user = env.getSource();
            String country = ...           // need to get "cf-ipcountry" from HTTP header here.
            return currencyService.getCurrency(user, country);
        });
    }
```
Thanks in advance!

**Comment by joostfarla** [comment, marked as answer]
You could provide contextual info as part of the execution input. For example:

```java
ExecutionInput executionInput = ExecutionInput.newExecutionInput()
    .query("{...your query...}")
    .graphQLContext(builder -> builder.put("cf-ipcountry", headerValue))
    .build();
```

In your datafetcher, you can then retrieve the context form the data fetching environment:

```java
public ConnectedDataFetcher<String> getCurrency() {
  return ConnectedDataFetcher.create(this.txManager, env -> {
    User user = env.getSource();
    String country = env.getGraphQlContext()
        .get("cf-ipcountry");
    return currencyService.getCurrency(user, country);
  });
}
```

You could also have a look at the Netflix DGS framework, which uses `graphql-java` under the hood and offers [a nice way to handle this](https://netflix.github.io/dgs/datafetching/#requestheader-requestparam-and-cookievalue).

**Reply by powerseed** [reply]
Sorry for the late reply! I adopted your solution but forgot to accept the answer.
Thank you so much!
